lib/wasi/src/syscalls/mod.rs

```diff
@@ -76,16 +75,18 @@ fn write_bytes<T: Write>(
 fn read_bytes<T: Read>(
     mut reader: T,
     memory: &Memory,
-    iovs_arr_cell: &[Cell<__wasi_iovec_t>],
+    iovs_arr_cell: Vec<WasmCell<__wasi_iovec_t>>,
 ) -> Result<u32, __wasi_errno_t> {
     let mut bytes_read = 0;
 
     for iov in iovs_arr_cell {
         let iov_inner = iov.get();
-        let bytes = WasmPtr::<u8, Array>::new(iov_inner.buf).deref(memory, 0, iov_inner.buf_len)?;
-        let mut raw_bytes: &mut [u8] =
-            unsafe { &mut *(bytes as *const [_] as *mut [_] as *mut [u8]) };
-        bytes_read += reader.read(raw_bytes).map_err(|_| __WASI_EIO)? as u32;
+        let mut raw_bytes: Vec<u8> = vec![0; iov_inner.buf_len as usize];
+        bytes_read += reader.read(&mut raw_bytes).map_err(|_| __WASI_EIO)? as u32;
+        memory
+            .view::<u8>()
+            .subarray(iov_inner.buf, iov_inner.buf + iov_inner.buf_len)
```

Comment:
This will now panic given inputs that it would not panic under before, we would just return the correct error before.

I think this is a bit of a regression in API usability

Reply:
To clarify: the issue is u32 overflow. This new API pushes it onto the caller when before we just handled it internally

Reply:
Fixed. I transformed them to `usize`